crates/iceberg/src/writer/file_writer/parquet_writer.rs

```diff
@@ -0,0 +1,390 @@
+}
+
+impl ParquetWriterBuilder {
+    /// To avoid EntiryTooSmall error, we set the minimum buffer size to 8MB if the given buffer size is smaller than it.
+    const MIN_BUFFER_SIZE: usize = 8 * 1024 * 1024;
+
+    /// Create a new `ParquetWriterBuilder`
+    /// To construct the write result, the schema should contain the `PARQUET_FIELD_ID_META_KEY` metadata for each field.
+    pub fn new(
+        init_buffer_size: usize,
+        props: WriterProperties,
+        schema: SchemaRef,
+        file_io: FileIO,
+        location_generator: Arc<dyn LocationGenerator>,
+    ) -> Self {
+        Self {
+            init_buffer_size,
+            props,
+            schema,
+            file_io,
+            location_generator,
+        }
+    }
+}
+
+impl FileWriterBuilder for ParquetWriterBuilder {
+    type R = ParquetWriter;
+
+    async fn build(self) -> crate::Result<Self::R> {
+        // Fetch field id from schema
+        let field_ids = self
+            .schema
+            .fields()
+            .iter()
+            .map(|field| {
+                field
+                    .metadata()
+                    .get(PARQUET_FIELD_ID_META_KEY)
+                    .ok_or_else(|| {
+                        Error::new(
+                            crate::ErrorKind::UnexpectedParameter,
+                            "field id not found in schema",
+                        )
+                    })?
+                    .parse::<i32>()
+                    .map_err(|err| {
+                        Error::new(
+                            crate::ErrorKind::UnexpectedParameter,
+                            "field id parse error",
+                        )
+                        .with_source(err)
+                    })
+            })
+            .collect::<crate::Result<Vec<_>>>()?;
+
+        let written_size = Arc::new(AtomicI64::new(0));
+        let file_path = self.location_generator.generate_name();
+        let out_file = self.file_io.new_output(file_path)?;
+        let inner_writer = TrackWriter::new(out_file.writer().await?, written_size.clone());
+        let init_buffer_size = if self.init_buffer_size < Self::MIN_BUFFER_SIZE {
+            Self::MIN_BUFFER_SIZE
+        } else {
+            self.init_buffer_size
+        };
+        let writer = AsyncArrowWriter::try_new(
+            inner_writer,
+            self.schema.clone(),
+            init_buffer_size,
+            Some(self.props),
+        )
+        .map_err(|err| {
+            Error::new(
+                crate::ErrorKind::Unexpected,
+                "build error from parquet writer",
+            )
+            .with_source(err)
+        })?;
+
+        Ok(ParquetWriter {
+            writer,
+            written_size,
+            current_row_num: 0,
+            out_file,
+            file_io: self.file_io,
+            field_ids,
+        })
+    }
+}
+
+/// `ParquetWriter`` is used to write arrow data into parquet file on storage.
+pub struct ParquetWriter {
```

Comment:
Is it possible to make `ParquetWriter<F: FileWriter>`? I think it's hidden under the `FileWriter` trait?

Reply:
I'm not sure what this means. ParquetWriter doesn't need another FileWriter. It's the lowest-level writer.

Reply:
Same as below.